[DirectX][NFC] Change specification of overload types and attribute in DXIL.td

bharadwajy opened the pull request:
- Specify overload types of DXIL Operation as list of types instead of a string.
 - Add supported DXIL type record definitions to `DXIL.td` leveraging `LLVMType` to avoid duplicate definitions.
 - Spell out DXIL Operation Attribute specification string.
 - Make corresponding changes to process the records in DXILEmitter.cpp

Commit:
[DirectX][NFC] Change specification of overload types and attributes in DXIL.td

 - Specify overload types of DXIL Operation as list of types instead of a string.
 - Add supported DXIL type record definitions to `DXIL.td` leveraging `LLVMType` to
   avoid duplicate definitions.
 - Spell out DXIL Operation Attribute specification string.
 - Make corresponding changes to process the records in DXILEmitter.cpp

bharadwajy commented:
@llvm-beanz @bogner @dmpots @python3kgae

python3kgae commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -268,36 +296,45 @@ static void emitDXILIntrinsicMap(std::vector<DXILOperationDesc> &Ops,
-      .Case("rn", "Attribute::ReadNone")
-      .Case("ro", "Attribute::ReadOnly")
+      .Case("NoReadMemory", "Attribute::ReadNone")
+      .Case("ReadMemory", "Attribute::ReadOnly")
```
Why not just ReadNone and ReadOnly?

bharadwajy replied:
> Why not just ReadNone and ReadOnly?

The thought was to emphasize *memory* read, but I am open to changing if the terms chosen don't sound palatable.

python3kgae replied:
ReadMemory lost the only part of ReadOnly, but ReadOnlyMemory feels bad too :(
Maybe OnlyReadMemory?

bharadwajy replied:
> ReadMemory lost the only part of ReadOnly, but ReadOnlyMemory feels bad too :( Maybe OnlyReadMemory?

How about `MemoryReadNone` and `MemoryReadOnly`?

dmpots replied:
Don't these map directly to the llvm function attributes? If so I think we should match those names (`readonly` and `readnone`).

bharadwajy replied:
> Don't these map directly to the llvm function attributes? 

Yes. 

> If so I think we should match those names (`readonly` and `readnone`). Using the decided naming convention (e.g. CamlCase).

As the `DXIL.td` is expected to be the eventual single source of reference for DXIL operations and affords declarative specification, the thought was to be a bit more descriptive than C++ sources, when possible.

I can change them as suggested.

bharadwajy replied:
I changed the type of `Attribute` to `EnumAttr` type to get the benefit of appropriately constrained values that can be specified and validating at compile-time as opposed to using a `string`.

Commit:
[DirectX][NFC] Change type of DXIL operation to EnumAttr in DXIL.td

Leverage EnumAttr class defined in llvm/IR/Attributes.td instead of
using an arbitrary string or defining a new one. This leverages
validation of acceptable attribute specification of DXIL operation
records added at compile-time.

dmpots commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -34,12 +35,42 @@ def BinaryUintCategory : DXILOpCategory<"Binary uint">;
 
+// ValueTypes specific to DXIL
+// Define Overload value type as an entity with no size and an arbitrary value
+// of 1024 - assuming that all currently defined values are less than 1024
```
What happens if LLVM changes to have a value >= 1024? Is it going to break in a easily recognizable way that is obvious to fix? How close are we to that limit?

bharadwajy replied:
> What happens if LLVM changes to have a value >= 1024? Is it going to break in a easily recognizable way that is obvious to fix? How close are we to that limit?

I have not investigated this. I am looking at a different way to define overload types without the reliance above as I share the same concerns. The current changes do not require these definitions. So, I have pushed a change that deletes these additions from this PR. Thanks!

dmpots commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -267,38 +296,51 @@ static void emitDXILIntrinsicMap(std::vector<DXILOperationDesc> &Ops,
+// @param Attr string reference
+// @return std::string Attribute enum string
+static std::string emitDXILOperationAttr(StringRef Attr) {
+  return StringSwitch<std::string>(Attr)
```
I see this is just following the existing code, but should we add some error handling for unexpected attributes here?

Commit:
Clean up unused type definitions in DXIL.td
